napari/utils/progress.py

```diff
@@ -0,0 +1,315 @@
+    """Adds ProgressBar to viewer Activity Dock and returns it.
+
+    Parameters
+    ----------
+    viewer_instance : qtViewer
+        current napari qtViewer instance
+
+    Returns
+    -------
+    ProgressBar
+        progress bar to associate with current iterable
+    """
+    pbar = ProgressBar(**kwargs)
+    viewer_instance.activityDock.widget().layout.addWidget(pbar)
+
+    return pbar
+
+
+def get_calling_function_name(max_depth: int):
+    """Inspect stack up to max_depth and return first function name outside of progress.py"""
+    for finfo in inspect.stack()[2:max_depth]:
+        if not finfo.filename.endswith("progress.py"):
+            return finfo.function
+
+    return None
+
+
+_tqdm_kwargs = {
+    p.name
+    for p in inspect.signature(tqdm.__init__).parameters.values()
+    if p.kind is not inspect.Parameter.VAR_KEYWORD and p.name != "self"
+}
+
+
+class progress(tqdm):
+    """This class inherits from tqdm and provides an interface for
+    progress bars in the napari viewer. Progress bars can be created
+    directly by wrapping an iterable or by providing a total number
+    of expected updates.
+
+    See tqdm.tqdm API for valid args and kwargs: https://tqdm.github.io/docs/tqdm/
+
+    Also, any keyword arguments to the :class:`ProgressBar` `QWidget`
+    are also accepted and will be passed to the ``ProgressBar``.
+
+    Examples
+    --------
+
+    >>> def long_running(steps=10, delay=0.1):
+    ...     for i in progress(range(steps)):
+    ...         sleep(delay)
+
+    it can also be used as a context manager:
+
+    >>> def long_running(steps=10, repeats=4, delay=0.1):
+    ...     with progress(range(steps)) as pbr:
+    ...         for i in pbr:
+    ...             sleep(delay)
+
+    or for manual updates:
+
+    >>> def manual_updates(total):
+    ...     pbr = progress(total=total)
+    ...     sleep(10)
+    ...     pbr.set_description("Step 1 Complete")
+    ...     pbr.update(1)
+
+    """
+
+    def __init__(
+        self,
+        iterable: Optional[Iterable] = None,
+        desc: Optional[str] = None,
+        total: Optional[int] = None,
+        *args,
+        **kwargs,
+    ) -> None:
+
+        # check if there's a napari viewer instance
+        viewer = get_viewer_instance()
+        if viewer is not None:
+            kwargs['gui'] = True
+
+        kwargs = kwargs.copy()
+        pbar_kwargs = {k: kwargs.pop(k) for k in set(kwargs) - _tqdm_kwargs}
+
+        super().__init__(iterable, desc, total, *args, **kwargs)
+        if viewer is None:
+            return
+        self.viewer = viewer
+
+        self._pbar = get_pbar(viewer, **pbar_kwargs)
+        if self.total is not None:
+            self._pbar.setRange(self.n, self.total)
+            self._pbar._set_value(self.n)
+        else:
+            self._pbar.setRange(0, 0)
+
+        if desc:
+            self.set_description(desc)
+        else:
+            desc = get_calling_function_name(max_depth=5)
```

Comment:
flagging this for @jni to comment on.  It _may_ be a bit heavy (take too much time) to inspect the frame stack just to provide the name of the calling function in the text of the progress bar?

Reply:
yeahh it does seem overkill and it won't be very informative if the user has nested progress bars or sets up multiple within a function anyway. I just wasn't sure what a sensible default would be...

Reply:
@tlambert03 I'm ok with it in this case — because as I understand it this happens once in front of a long-running operation? Rather than e.g. once per layer-add. Having said that I'm totally happy to just use a generic name like "progress..." if no name is given.

Reply:
Sounds good!

Reply:
This is going to produce total garbage when using `progress` with `thread_worker`, so we do need to come up with something better if we want to introspect - otherwise we just use some default... Although I think maybe we should be distinguishing between the progress bar name and its description.

Reply:
I've removed this for now, we can keep thinking about a better default.